src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -34,46 +36,52 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     // Boolean to update User's main page 
     Boolean showNewTab = false;
 
+    User user;
     if (userService.isUserLoggedIn()) {
-      userEmail = userService.getCurrentUser().getEmail();
       userId = userService.getCurrentUser().getUserId();
+      userEmail = userService.getCurrentUser().getEmail();
       logoutUrl = userService.createLogoutURL(urlToRedirect);
       showNewTab = true;
     } else {
       loginUrl = userService.createLoginURL(urlToRedirect);
     }
 
-    User user = new User(userId, loginUrl, logoutUrl, userEmail, showNewTab);
-    
-    if (userId != "null") {
-        checkUserInDatastore(user);
+    user = new User(userId, userEmail);
+
+    if ((userId != "null") && (!isUserInDatastore(user))) {
+      storeUserToDatastore(user);
     }
 
+    // Aggregate response information
+    Map<String, Object> jsonInfo = new HashMap<>();
+    jsonInfo.put("userInfo", user);
+    jsonInfo.put("showTabStatus", showNewTab);
+    jsonInfo.put("logoutUrl", logoutUrl);
+    jsonInfo.put("loginUrl", loginUrl);
+
     response.setContentType("application/json;");
-    response.getWriter().println(new Gson().toJson(user));
+    response.getWriter().println(new Gson().toJson(jsonInfo));
   }
 
-  public void checkUserInDatastore(User user) {
-    // ** NEED TO WRITE TEST FOR BEHAVIORAL TESTING
+  public boolean isUserInDatastore(User user) {
 
-    Key userKey = KeyFactory.createKey("User", user.getUserId());
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Key userKey = KeyFactory.createKey("User", user.getUserId());
     Entity userEntity;
 
     // Check if user is in datastore
     try {
       userEntity = datastore.get(userKey);
```

Comment:
is userEntity being used anywhere?

Reply:
It's used to check if the userEntity is in the datastore. If not, we create one

Reply:
If I am reading the documentation correctly, datastore.get  returns "null" if the object doesn't exist. The datastore exception only happens if something else goes wrong (server timeout, for example). Which documentation are you looking at for tat says this will throw "EntittyNotFoundException"?

Reply:
Never mind. I was looking at different version of the Client library..